sway/config/output.c

```diff
@@ -527,24 +527,23 @@ static bool finalize_output_config(struct output_config *oc, struct sway_output
 		return true;
 	}
 
-	if (oc) {
-		enum scale_filter_mode scale_filter_old = output->scale_filter;
-		switch (oc->scale_filter) {
-			case SCALE_FILTER_DEFAULT:
-			case SCALE_FILTER_SMART:
-				output->scale_filter = ceilf(wlr_output->scale) == wlr_output->scale ?
-					SCALE_FILTER_NEAREST : SCALE_FILTER_LINEAR;
-				break;
-			case SCALE_FILTER_LINEAR:
-			case SCALE_FILTER_NEAREST:
-				output->scale_filter = oc->scale_filter;
-				break;
-		}
-		if (scale_filter_old != output->scale_filter) {
-			sway_log(SWAY_DEBUG, "Set %s scale_filter to %s", oc->name,
-				sway_output_scale_filter_to_string(output->scale_filter));
-			wlr_damage_ring_add_whole(&output->scene_output->damage_ring);
-		}
+	enum scale_filter_mode scale_filter_old = output->scale_filter;
+	enum scale_filter_mode scale_filter_new = oc ? oc->scale_filter : SCALE_FILTER_DEFAULT;
+	switch (scale_filter_new) {
+		case SCALE_FILTER_DEFAULT:
+		case SCALE_FILTER_SMART:
+			output->scale_filter = ceilf(wlr_output->scale) == wlr_output->scale ?
+				SCALE_FILTER_NEAREST : SCALE_FILTER_LINEAR;
+			break;
+		case SCALE_FILTER_LINEAR:
+		case SCALE_FILTER_NEAREST:
+			output->scale_filter = scale_filter_new;
+			break;
+	}
+	if (scale_filter_old != output->scale_filter) {
+		sway_log(SWAY_DEBUG, "Set %s scale_filter to %s", oc->name,
+			sway_output_scale_filter_to_string(output->scale_filter));
+		wlr_damage_ring_add_whole(&output->scene_output->damage_ring);
```

Comment:
Bypassing scene and applying damage like this isn't ideal. The internal pending_damage thing also needs to be updated ideally. We aren't communicating this damage with the backend. But the scale filtering is already a hack...

Reply:
Yeah, it's not great - that part is still carried over from the original code though. It just got moved to this "finalize" helper when modeset got introduced, and now got a slight change to be more stable. :/

Reply:
I think I originally authored that. We might need to take a page out of labwc and vendor a damage function.